Improved brigadier documentation (#436)
PaperMC/docs: docs/paper/dev/api/commands-api/arguments.mdx

---

> slug: /dev/commands-api/arguments

Reviewer: Not sure that I like these slugs. I prefer either `/commands/` or `/command-api/` as the former would be `/commands/commands`

> description: A guide to Paper's Brigadier command API.

Reviewer: Not specific to the page

Author: fixed

---

# Arguments

> Argumenttypes are datatypes that we can use instead of strings.

Reviewer: ```suggestion
Argument types are datatypes that we can use instead of strings.
```

---
slug: /dev/commands-api/arguments
description: A guide to Paper's Brigadier command API.
---

# Arguments

Argumenttypes are datatypes that we can use instead of strings.

:::danger[Experimental]

Paper's command system is still experimental and may change in the future.

:::

> ## Basic usage of arguments

Reviewer 2: ```suggestion
## Basic usage
```

                            );
                            return Command.SINGLE_SUCCESS;
                        })
                ).build()
            );
        });
    }
}
```
This command has one argument of the `Enchantment` datatype. The advantage of using
Brigadier over the `TabExecutor` interface is that you don't have to convert the
string that is passed to the `CommandExecutor#onCommand()` function as an argument to 
an enchantment in this case because `ctx.getArgument("enchantmentargument", Enchantment.class)`
returns a value of the `Enchantment` datatype. Additionally, you don't have to write 
error handling functionality yourself as the client checks whether the argument the
user wrote in the chat is one of the `minecraft:enchantment` type and if it isn't,
the user will get an error. Another advantage `ArgumentTypes` have over string-based argument systems like
Bukkit's `TabExecutor` is that we can work with our own types and the completions list 
is not always sorted alphanumerically. This may sound insignificant but when you want an input 
of the type integer and suggest the numbers 0-10, for example, your list will look like this because

> it needs to be converted to strings:`0, 1, 10, 2, 3, ...`. Looks awful, right?

Reviewer 2: this paragraph is quite complex, the long sentences could be broken up into something more easily digestible

> ## Enchantment types

Reviewer 2: ```suggestion
## Types
```

Reviewer: Not addressed

> By default, you can use [registries](../registries) to get simple argument types like

Reviewer 2: ```suggestion
By default, you can use [the registry API](../registries.mdx) to get simple argument types like
```

Reviewer: Not addressed

> blocks, items, potions and many more. In the example above, we used the Enchantment 
> Argument type but there are many others:

Reviewer 2: ```suggestion
blocks, items, potions and many more. In the example above, we used the `Enchantment` 
argument type, but there are many others:
```

Reviewer: Not addressed

### Predefined types (Registry)

| Registry key name   | Return datatype class | Description                                                                                                                                                                                                                                                                                  |
|---------------------|-----------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| GAME_EVENT          | GameEvent             | Events in the game (eating, flying with an elytra etc.)                                                                                                                                                                                                                                      |

> | STRUCTURE_TYPE      | StructureType         | [Structures](https://minecraft.wiki/w/Structure#Overworld) - Beware that, as of Paper 1.20.6-147, there is also a Registry named `STRUCTURE` that is not recognized by the client and makes it unable to connect. This has been reported to Paper developers and will be fixed in a future release. |

Reviewer 2: presuming this is referring to https://github.com/PaperMC/Paper/issues/10991; this bug hasn't been triaged yet and I don't think it needs to be documented here, as it will likely be promptly fixed and the comment obsoleted

| MEMORY_MODULE_TYPE  | MemoryKey             | Keys for saving per-entity data                                                                                                                                                                                                                                                              |
| FROG_VARIANT        | Frog.Variant          | [Frog variants](https://minecraft.wiki/w/Frog)                                                                                                                                                                                                                                               |
| MAP_DECORATION_TYPE | MapCursor.Type        | [Types of sprites displayed on a map](https://minecraft.wiki/w/Map#Map_icons)                                                                                                                                                                                                                |
| FLUID               | Fluid                 | [Fluid types](https://minecraft.wiki/w/Fluid)     

> ### Predefined types (in Brigadier itself)

Reviewer 2: this section can probably be merged into the one above, not a fan of the title repetition

Author: Are you talking about the description?

> Brigadier itself also specifies many argument types. For more information on them, see <Javadoc name={"io.papermc.paper.command.brigadier.argument.ArgumentTypes"}>LifecycleEventManager</Javadoc>

Reviewer 2: ```suggestion
Brigadier itself also specifies many argument types. For more information on them, see <Javadoc name={"io.papermc.paper.command.brigadier.argument.ArgumentTypes"}>ArgumentTypes</Javadoc>.
```

Reviewer: Not addressed

### Custom types

Custom arguments can be created by implementing the <Javadoc name={"io.papermc.paper.command.brigadier.argument.CustomArgumentType"}>CustomArgumentType</Javadoc> 
interface. 

> Now, lets say that we want to implement a command which lets you order ice cream. For that,

Reviewer 2: ```suggestion
Now, let's say that we want to implement a command which lets you order ice cream. For that,
```

Reviewer: Not addressed

        );
    }
}
```

That's a lot of code, so let's start from the top. We implemented the `CustomArgumentType.Converted`
interface. This interface takes two type arguments: our custom enum, T, and a type that is native to 
Minecraft, such as String, Integer, etc., called N. The native type exists so that the client can use
the input data, as it doesn't know what our custom `IceCreamType` is.

The first method called `convert()` converts the native type (in this case `String`)
into our own custom type called `IceCreamType`, as the name suggests. This is so we 
can keep using our custom type internally. If we wouldn't do that, there would be no point in using a 
custom type. If the specified input isn't found in our enum of available flavors, we
output an error.

The second method called `getNativeType()` returns the type of string that our command argument uses.
In our case, we're using a single word, so we return `StringArgumentType.word()`.

In the last method, `listSuggestions()`, we return `CompletableFuture<Suggestions>` so that the client
can suggest all available options. We can even add tooltips to the suggestions to explain them in greater
detail. 

> As a last step, we need to register the Command:

Reviewer 2: ```suggestion
As a last step, we need to register the command:
```

```java
public void onEnable() {
    LifecycleEventManager manager = this.getLifecycleManager();
    manager.registerEventHandler(LifecycleEvents.COMMANDS, event -> {
        final Commands commands = event.registrar();
        commands.register(Commands.literal("ordericecream")
            .then(
                Commands.argument("flavor", new IceCreamTypeArgument()).executes((commandContext -> {
                    IceCreamType argumentResponse = commandContext.getArgument("flavor", IceCreamType.class); // Gets the raw argument
                    commandContext.getSource().getSender().sendMessage(Component.text("You ordered: " + argumentResponse));
                    return 1;
                }))
            ).build()
        );
    });
}
```

> Now that we have registered the command, we can execute it ingame:

Reviewer 2: ```suggestion
Now that we have registered the command, we can execute it in-game:
```

![command with suggestions](./assets/icecreamargument.gif)

> Look, we can even see our tooltip and if we execute the command, we get the message we specified

Reviewer 2: ```suggestion
Look, we can even see our tooltip and if we execute the command, we get the message we specified:
```
